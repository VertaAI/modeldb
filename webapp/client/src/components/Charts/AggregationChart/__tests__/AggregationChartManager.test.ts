import {
  makeShallowRenderer,
  findByDataTestAttribute,
} from 'core/shared/utils/tests/react/helpers';
import {
  flatData_allMetricHyp,
  flatData_someMetricMissing,
  flatData_someHypMissing,
  flatData_onlyHyperparams,
  flatData_onlyMetric,
} from '../../__tests__/mocks';
import {
  IAggregationChartProps,
  AggregationChartManager,
} from '../AggregationChartManager';
import {
  mockAggChartConfig,
  mockNoMetricAggChartConfig,
  mockNoHypAggChartConfig,
  mockHyperparamKeys,
  mockNoHyperparamKeys,
  mockMetricKeys,
  mockNoMetricKeys,
} from './mocks';

const makeShallowComponent = (props: Partial<IAggregationChartProps> = {}) => {
  const defaultProps: IAggregationChartProps = {
    aggregationChartConfig: {
      selectedMetric: '',
      selectedHyperparameter: '',
      selectedAggregationType: '',
      selectedChartType: '',
    },
    metricKeys: new Set(['']),
    hyperparameterKeys: new Set(['']),
    genericChartData: [],
    updateAggregationChartConfig: jest.fn(),
  };
  return {
    component: makeShallowRenderer<IAggregationChartProps>(
      AggregationChartManager,
      defaultProps
    )(props),
  };
};

describe('(aggregation chart)', () => {
  describe('(data: all rows with metric and hyp)', () => {
    const props = {
      aggregationChartConfig: mockAggChartConfig,
      metricKeys: mockMetricKeys,
      hyperparameterKeys: mockHyperparamKeys,
      genericChartData: flatData_allMetricHyp,
    };
    const { component } = makeShallowComponent(props);

    it('should load  a valid aggregationChart component', () => {
      const wrapper = findByDataTestAttribute('aggregation-chart', component);
      expect(wrapper.length).toBe(1);
    });

    it('aggregationChart config loaded and its value as expected', () => {
      const classInstance = component.instance() as AggregationChartManager;
      const selectedMetric = classInstance.state.userSelectedMetricVal;
      expect(selectedMetric).toBe('val_acc');

      const selectedHyperparam =
        classInstance.state.userSelectedHyperparameterVal;
      expect(selectedHyperparam).toBe('C');

      const selectedAggregationType =
        classInstance.state.userSelectedAggregationType;
      expect(selectedAggregationType).toBe('average');

      const selectedChartType = classInstance.state.userSelectedChartType;
      expect(selectedChartType).toBe('bar-chart');
    });

    it('should render aggregationChart component that matches snapshot', () => {
      expect(
        findByDataTestAttribute('aggregation-chart', component)
      ).toMatchSnapshot();
    });
  });

  describe('(data: some metric missing)', () => {
    const props = {
      aggregationChartConfig: mockAggChartConfig,
      metricKeys: mockMetricKeys,
      hyperparameterKeys: mockHyperparamKeys,
      genericChartData: flatData_someMetricMissing,
    };
    const { component } = makeShallowComponent(props);

    it('should load  a valid aggregationChart component', () => {
      const wrapper = findByDataTestAttribute('aggregation-chart', component);
      expect(wrapper.length).toBe(1);
    });

    it('aggregationChart config loaded and its value as expected', () => {
      const classInstance = component.instance() as AggregationChartManager;
      const selectedMetric = classInstance.state.userSelectedMetricVal;
      expect(selectedMetric).toBe('val_acc');

      const selectedHyperparam =
        classInstance.state.userSelectedHyperparameterVal;
      expect(selectedHyperparam).toBe('C');

      const selectedAggregationType =
        classInstance.state.userSelectedAggregationType;
      expect(selectedAggregationType).toBe('average');

      const selectedChartType = classInstance.state.userSelectedChartType;
      expect(selectedChartType).toBe('bar-chart');
    });

    it('should render aggregationChart component that matches snapshot', () => {
      expect(
        findByDataTestAttribute('aggregation-chart', component)
      ).toMatchSnapshot();
    });
  });

  describe('(data: some hyp missing)', () => {
    const props = {
      aggregationChartConfig: mockAggChartConfig,
      metricKeys: mockMetricKeys,
      hyperparameterKeys: mockHyperparamKeys,
      genericChartData: flatData_someHypMissing,
    };
    const { component } = makeShallowComponent(props);

    it('should load  a valid aggregationChart component', () => {
      const wrapper = findByDataTestAttribute('aggregation-chart', component);
      expect(wrapper.length).toBe(1);
    });

    it('aggregationChart config loaded and its value as expected', () => {
      const classInstance = component.instance() as AggregationChartManager;
      const selectedMetric = classInstance.state.userSelectedMetricVal;
      expect(selectedMetric).toBe('val_acc');

      const selectedHyperparam =
        classInstance.state.userSelectedHyperparameterVal;
      expect(selectedHyperparam).toBe('C');

      const selectedAggregationType =
        classInstance.state.userSelectedAggregationType;
      expect(selectedAggregationType).toBe('average');

      const selectedChartType = classInstance.state.userSelectedChartType;
      expect(selectedChartType).toBe('bar-chart');
    });

    it('should render aggregationChart component that matches snapshot', () => {
      expect(
        findByDataTestAttribute('aggregation-chart', component)
      ).toMatchSnapshot();
    });
  });

  describe('(data: only hyperparams)', () => {
    const props = {
      aggregationChartConfig: mockNoMetricAggChartConfig,
      metricKeys: mockNoMetricKeys,
      hyperparameterKeys: mockHyperparamKeys,
      genericChartData: flatData_onlyHyperparams,
    };
    const { component } = makeShallowComponent(props);
    it('should load  a valid aggregationChart component', () => {
      const wrapper = findByDataTestAttribute('aggregation-chart', component);
      expect(wrapper.length).toBe(1);
    });

    it('aggregationChart config loaded and its value as expected', () => {
      const classInstance = component.instance() as AggregationChartManager;
      const selectedMetric = classInstance.state.userSelectedMetricVal;
      expect(selectedMetric).toBe('data not available');

      const selectedHyperparam =
        classInstance.state.userSelectedHyperparameterVal;
      expect(selectedHyperparam).toBe('C');

      const selectedAggregationType =
        classInstance.state.userSelectedAggregationType;
      expect(selectedAggregationType).toBe('average');

      const selectedChartType = classInstance.state.userSelectedChartType;
      expect(selectedChartType).toBe('bar-chart');
    });

    it('should render aggregationChart component that matches snapshot', () => {
      expect(
        findByDataTestAttribute('aggregation-chart', component)
      ).toMatchSnapshot();
    });
  });

  describe('(data: only hyperparams)', () => {
    const props = {
      aggregationChartConfig: mockNoHypAggChartConfig,
      metricKeys: mockMetricKeys,
      hyperparameterKeys: mockNoHyperparamKeys,
      genericChartData: flatData_onlyMetric,
    };
    const { component } = makeShallowComponent(props);

    it('should load  a valid aggregationChart component', () => {
      const wrapper = findByDataTestAttribute('aggregation-chart', component);
      expect(wrapper.length).toBe(1);
    });

    it('aggregationChart config loaded and its value as expected', () => {
      const classInstance = component.instance() as AggregationChartManager;
      const selectedMetric = classInstance.state.userSelectedMetricVal;
      expect(selectedMetric).toBe('val_acc');

      const selectedHyperparam =
        classInstance.state.userSelectedHyperparameterVal;
      expect(selectedHyperparam).toBe('data not available');

      const selectedAggregationType =
        classInstance.state.userSelectedAggregationType;
      expect(selectedAggregationType).toBe('average');

      const selectedChartType = classInstance.state.userSelectedChartType;
      expect(selectedChartType).toBe('bar-chart');
    });

    it('should render aggregationChart component that matches snapshot', () => {
      expect(
        findByDataTestAttribute('aggregation-chart', component)
      ).toMatchSnapshot();
    });
  });
});
