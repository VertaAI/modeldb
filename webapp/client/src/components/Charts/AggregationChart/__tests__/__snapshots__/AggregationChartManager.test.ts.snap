// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(aggregation chart) (data: all rows with metric and hyp) should render aggregationChart component that matches snapshot 1`] = `
<div
  className="chart_section_wrapper"
  data-test="aggregation-chart"
>
  <div
    className="chart_header"
  >
    Aggregated Metrics/Hyperparameters Chart
  </div>
  <div
    className="chart_config_selectors"
  >
    <ChartConfigDropdown
      label="Metric :"
      onChange={[Function]}
      options={
        Set {
          "val_acc",
        }
      }
      value="val_acc"
    />
    <ChartConfigDropdown
      label="Hyperparameter :"
      onChange={[Function]}
      options={
        Set {
          "C",
          "max_iter",
          "solver",
        }
      }
      value="C"
    />
    <ChartConfigDropdown
      isDisabled={false}
      label="Aggregation Types :"
      onChange={[Function]}
      options={
        Array [
          "average",
          "sum",
          "median",
          "variance",
          "stdev",
          "count",
        ]
      }
      value="average"
    />
    <ChartConfigDropdown
      label="Chart Type :"
      onChange={[Function]}
      options={
        Array [
          "bar-chart",
          "box-plot",
        ]
      }
      value="bar-chart"
    />
  </div>
  <div>
    <BarChart
      data={
        Array [
          Object {
            "key": 0.000001,
            "value": 0.81955,
          },
          Object {
            "key": 0.0001,
            "value": 0.8012,
          },
        ]
      }
      xLabel="C"
      yLabel="val_acc"
    />
  </div>
</div>
`;

exports[`(aggregation chart) (data: only hyperparams) should render aggregationChart component that matches snapshot 1`] = `
<div
  className="chart_section_wrapper"
  data-test="aggregation-chart"
>
  <div
    className="chart_header"
  >
    Aggregated Metrics/Hyperparameters Chart
  </div>
  <div
    className="chart_config_selectors"
  >
    <ChartConfigDropdown
      label="Metric :"
      onChange={[Function]}
      options={Set {}}
      value="data not available"
    />
    <ChartConfigDropdown
      label="Hyperparameter :"
      onChange={[Function]}
      options={
        Set {
          "C",
          "max_iter",
          "solver",
        }
      }
      value="C"
    />
    <ChartConfigDropdown
      isDisabled={false}
      label="Aggregation Types :"
      onChange={[Function]}
      options={
        Array [
          "average",
          "sum",
          "median",
          "variance",
          "stdev",
          "count",
        ]
      }
      value="average"
    />
    <ChartConfigDropdown
      label="Chart Type :"
      onChange={[Function]}
      options={
        Array [
          "bar-chart",
          "box-plot",
        ]
      }
      value="bar-chart"
    />
  </div>
  <div>
    <BarChart
      data={Array []}
      xLabel="C"
      yLabel="data not available"
    />
  </div>
</div>
`;

exports[`(aggregation chart) (data: only hyperparams) should render aggregationChart component that matches snapshot 2`] = `
<div
  className="chart_section_wrapper"
  data-test="aggregation-chart"
>
  <div
    className="chart_header"
  >
    Aggregated Metrics/Hyperparameters Chart
  </div>
  <div
    className="chart_config_selectors"
  >
    <ChartConfigDropdown
      label="Metric :"
      onChange={[Function]}
      options={
        Set {
          "val_acc",
        }
      }
      value="val_acc"
    />
    <ChartConfigDropdown
      label="Hyperparameter :"
      onChange={[Function]}
      options={Set {}}
      value="data not available"
    />
    <ChartConfigDropdown
      isDisabled={false}
      label="Aggregation Types :"
      onChange={[Function]}
      options={
        Array [
          "average",
          "sum",
          "median",
          "variance",
          "stdev",
          "count",
        ]
      }
      value="average"
    />
    <ChartConfigDropdown
      label="Chart Type :"
      onChange={[Function]}
      options={
        Array [
          "bar-chart",
          "box-plot",
        ]
      }
      value="bar-chart"
    />
  </div>
  <div>
    <BarChart
      data={Array []}
      xLabel="data not available"
      yLabel="val_acc"
    />
    <div
      className="missing_count_block"
    >
      missing data count 
      4
       - [
      data not available
      ]
    </div>
  </div>
</div>
`;

exports[`(aggregation chart) (data: some hyp missing) should render aggregationChart component that matches snapshot 1`] = `
<div
  className="chart_section_wrapper"
  data-test="aggregation-chart"
>
  <div
    className="chart_header"
  >
    Aggregated Metrics/Hyperparameters Chart
  </div>
  <div
    className="chart_config_selectors"
  >
    <ChartConfigDropdown
      label="Metric :"
      onChange={[Function]}
      options={
        Set {
          "val_acc",
        }
      }
      value="val_acc"
    />
    <ChartConfigDropdown
      label="Hyperparameter :"
      onChange={[Function]}
      options={
        Set {
          "C",
          "max_iter",
          "solver",
        }
      }
      value="C"
    />
    <ChartConfigDropdown
      isDisabled={false}
      label="Aggregation Types :"
      onChange={[Function]}
      options={
        Array [
          "average",
          "sum",
          "median",
          "variance",
          "stdev",
          "count",
        ]
      }
      value="average"
    />
    <ChartConfigDropdown
      label="Chart Type :"
      onChange={[Function]}
      options={
        Array [
          "bar-chart",
          "box-plot",
        ]
      }
      value="bar-chart"
    />
  </div>
  <div>
    <BarChart
      data={
        Array [
          Object {
            "key": 0.000001,
            "value": 0.81955,
          },
        ]
      }
      xLabel="C"
      yLabel="val_acc"
    />
    <div
      className="missing_count_block"
    >
      missing data count 
      2
       - [
      C
      ]
    </div>
  </div>
</div>
`;

exports[`(aggregation chart) (data: some metric missing) should render aggregationChart component that matches snapshot 1`] = `
<div
  className="chart_section_wrapper"
  data-test="aggregation-chart"
>
  <div
    className="chart_header"
  >
    Aggregated Metrics/Hyperparameters Chart
  </div>
  <div
    className="chart_config_selectors"
  >
    <ChartConfigDropdown
      label="Metric :"
      onChange={[Function]}
      options={
        Set {
          "val_acc",
        }
      }
      value="val_acc"
    />
    <ChartConfigDropdown
      label="Hyperparameter :"
      onChange={[Function]}
      options={
        Set {
          "C",
          "max_iter",
          "solver",
        }
      }
      value="C"
    />
    <ChartConfigDropdown
      isDisabled={false}
      label="Aggregation Types :"
      onChange={[Function]}
      options={
        Array [
          "average",
          "sum",
          "median",
          "variance",
          "stdev",
          "count",
        ]
      }
      value="average"
    />
    <ChartConfigDropdown
      label="Chart Type :"
      onChange={[Function]}
      options={
        Array [
          "bar-chart",
          "box-plot",
        ]
      }
      value="bar-chart"
    />
  </div>
  <div>
    <BarChart
      data={
        Array [
          Object {
            "key": 0.000001,
            "value": 0.81955,
          },
        ]
      }
      xLabel="C"
      yLabel="val_acc"
    />
  </div>
</div>
`;
